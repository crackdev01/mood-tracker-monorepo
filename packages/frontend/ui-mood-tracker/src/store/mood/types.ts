// Define Mood Store State.
export interface MoodState {
  mood: [];
}

// Define Mood Store actions.
export enum MoodActions {
  FETCH_MOODS = 'FETCH_MOODS',
  RENDER_MOODS = 'RENDER_MOODS',
}
