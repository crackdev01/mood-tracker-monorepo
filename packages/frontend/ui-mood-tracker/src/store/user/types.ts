export interface UserRequest {
  username: string;
  password: string;
}

export interface Jwt {
  exp: number;
  iat: number;
  username: string;
  uuid: string;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface User {
  accessToken: string;
  decodedAccessToken: Jwt | null;
  // We set the coordinate values if user has provided permission.
  // Else we set to false.
  location?: Coordinates | string;
}

// Define User Store State.
export interface UserState {
  user: User;
}

// Define User Store actions.
export enum UserActions {
  FETCH_USER = 'FETCH_USER',
  RENDER_USER = 'RENDER_USER',
  LOGOUT_USER = 'LOGOUT_USER',
  SET_LOCATION_SUCCESS = 'SET_LOCATION_SUCCESS',
  SET_LOCATION_DENIED = 'SET_LOCATION_DENIED',
  SET_LOCATION_ERROR = 'SET_LOCATION_ERROR',
}
