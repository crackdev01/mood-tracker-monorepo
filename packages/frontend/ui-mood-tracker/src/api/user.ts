import { moodTrackerApi } from './core';

export async function authenticateUser(): Promise<any> {
  return await moodTrackerApi.post('/auth/login', { username: 'testuser', password: 'password' });
}
