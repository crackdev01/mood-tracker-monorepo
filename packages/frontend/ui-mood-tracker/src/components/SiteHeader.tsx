import React from 'react';
import { useTranslation } from 'react-i18next';
import { Link } from 'react-router-dom';
import { Menu } from 'semantic-ui-react';

import '../App.scss';

function SiteHeader() {
  const { t } = useTranslation(['SiteHeader']);

  return (
    <Menu pointing secondary>
      <Menu.Item header>
        <Link to="login">{t('header')}</Link>
      </Menu.Item>
      <Menu.Item>
        <Link to="/mood-entry">{t('moodEntry')}</Link>
      </Menu.Item>
      <Menu.Item>
        <Link to="/statistics">{t('statistics')}</Link>
      </Menu.Item>
    </Menu>
  );
}

export default SiteHeader;
