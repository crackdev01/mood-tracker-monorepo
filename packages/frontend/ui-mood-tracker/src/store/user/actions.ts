import { UserRequest, UserActions } from './types';

// FIXME: This should not require any payload.
export const fetchUser = (payload: UserRequest) => {
  return {
    type: UserActions.FETCH_USER,
    payload,
  };
};

export const renderUser = (payload: any) => {
  return {
    type: UserActions.RENDER_USER,
    payload,
  };
};
