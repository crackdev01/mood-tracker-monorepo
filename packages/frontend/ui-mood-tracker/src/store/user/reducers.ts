import { userState } from './state';
import { UserActions } from './types';

export const user = (state = userState, action: any) => {
  switch (action.type) {
    case UserActions.FETCH_USER:
      return {
        ...state,
      };
    case UserActions.RENDER_USER:
      return {
        ...state,
        user: action.data,
      };
    default:
      return state;
  }
};
