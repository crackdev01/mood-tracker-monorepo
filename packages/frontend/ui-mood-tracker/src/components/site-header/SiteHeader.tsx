import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { Link, useLocation } from 'react-router-dom';
import { Dropdown, Header, Menu } from 'semantic-ui-react';

import { UserActions } from '../../store/user/types';
import { ApplicationState } from '../../store';

import './site-header.scss';

enum LocaleEnum {
  English = 'en',
  Deutsche = 'de',
}

const SiteHeader = () => {
  const dispatch = useDispatch();
  const { t, i18n } = useTranslation(['SiteHeader']);
  const user = useSelector((state: ApplicationState) => state.userReducer.user);
  const [currentUser, setCurrentUser] = useState('');
  const location = useLocation();

  const isAuthenticated = user ? !!user.accessToken : false;

  const updateLanguage = () => {
    const { language } = i18n;
    i18n.changeLanguage(language === LocaleEnum.English ? LocaleEnum.Deutsche : LocaleEnum.English);
  };

  const logout = () => {
    dispatch({
      type: UserActions.LOGOUT_USER,
    });
  };

  useEffect(() => {
    const username = user.decodedAccessToken ? user.decodedAccessToken.username : '';
    setCurrentUser(username);
  }, [user]);

  return (
    <Menu pointing secondary className="site-header">
      <Menu.Item header>
        <Header as="h3">{t('header')}</Header>
      </Menu.Item>
      {isAuthenticated && (
        <Menu.Item active={location.pathname === '/mood-entry' || location.pathname === '/'}>
          <Link to="/mood-entry">{t('moodEntry')}</Link>
        </Menu.Item>
      )}
      {isAuthenticated && (
        <Menu.Item active={location.pathname === '/statistics'}>
          <Link id="statistics" to="/statistics">
            {t('statistics')}
          </Link>
        </Menu.Item>
      )}
      {isAuthenticated && (
        <Menu.Menu position="right">
          <Dropdown id="toggle-menu" item text={currentUser}>
            <Dropdown.Menu>
              <Dropdown.Item>{t('menu.actions.location')}</Dropdown.Item>
              <Dropdown.Item id="change-language" onClick={updateLanguage}>
                {t('menu.actions.changeLanguage')}
              </Dropdown.Item>
              <Dropdown.Item id="logout" onClick={logout}>
                {t('menu.actions.logout')}
              </Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </Menu.Menu>
      )}
    </Menu>
  );
};

export default SiteHeader;
