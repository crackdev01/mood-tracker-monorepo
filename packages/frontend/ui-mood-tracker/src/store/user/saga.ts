import { all, call, put, takeEvery, fork } from 'redux-saga/effects';
import jwt_decode from 'jwt-decode';

import { User, UserActions } from './types';
import { userState } from './state';
import { authenticateUser } from '../../api/user';

export function* loginUserCall(action: any) {
  const { data } = yield call(authenticateUser, action.payload);
  const user: User = {
    accessToken: data.accessToken,
    decodedAccessToken: jwt_decode(data.accessToken),
  };
  yield put({ type: UserActions.RENDER_USER, data: user });
}

export function* loginUser() {
  yield takeEvery(UserActions.FETCH_USER, loginUserCall);
}

export function* setUserLocationSuccessCall(action: any) {
  const { location } = action.payload;
  yield put({ type: UserActions.RENDER_USER, data: location });
}

export function* setUserLocationSuccess() {
  yield takeEvery(UserActions.SET_LOCATION_SUCCESS, setUserLocationSuccessCall);
}

export function* setUserLocationErrorCall() {
  const location = UserActions.SET_LOCATION_ERROR;
  yield put({ type: UserActions.RENDER_USER, data: location });
}

export function* setUserLocationError() {
  yield takeEvery(UserActions.SET_LOCATION_ERROR, setUserLocationErrorCall);
}

export function* setUserLocationDeniedCall() {
  const location = UserActions.SET_LOCATION_DENIED;
  yield put({ type: UserActions.RENDER_USER, data: location });
}

export function* setUserLocationDenied() {
  yield takeEvery(UserActions.SET_LOCATION_DENIED, setUserLocationDeniedCall);
}

export function* logoutUserCall() {
  // Reset use store to default state to trigger logout.
  yield put({ type: UserActions.RENDER_USER, data: userState });
}

export function* logoutUser() {
  yield takeEvery(UserActions.LOGOUT_USER, logoutUserCall);
}

export function* userSaga() {
  yield all([fork(loginUser), fork(logoutUser)]);
}

export default userSaga;
