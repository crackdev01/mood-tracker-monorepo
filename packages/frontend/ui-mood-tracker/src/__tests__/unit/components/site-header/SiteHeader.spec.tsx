import React from 'react';
import { useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import SiteHeader from '../../../../components/site-header/SiteHeader';

import { mockUser } from '../../../fixtures';
import { UserActions } from '../../../../store/user/types';

Enzyme.configure({ adapter: new Adapter() });

const useSelectorMock = useSelector as jest.Mock<any>;
const useLocationMock = useLocation as jest.Mock<any>;
const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch,
  useSelector: jest.fn(),
}));
jest.mock('react-router-dom', () => ({
  Link: jest.fn(),
  useLocation: jest.fn(),
}));

describe('SiteHeader', () => {
  beforeEach(mockDispatch.mockReturnValueOnce({}));

  describe('renders', () => {
    test('only displays header for unauthenticated user', () => {
      useSelectorMock.mockImplementationOnce((s) =>
        s({ userReducer: { user: { accessToken: null } } })
      );
      useLocationMock.mockReturnValueOnce(() => ({
        location: {
          pathname: '/mood-entry',
        },
      }));
      const w = shallow(<SiteHeader />);
      expect(w.debug()).toMatchSnapshot();
    });

    test('displays header entries for authenticated user', () => {
      useSelectorMock.mockImplementationOnce((s) => s({ userReducer: { user: mockUser } }));
      useLocationMock.mockReturnValueOnce({
        location: {
          pathname: '/mood-entry',
        },
      });
      const w = shallow(<SiteHeader />);
      expect(w.debug()).toMatchSnapshot();
      // FIXME: snapshot doesn't have user info.
    });

    test('changes language on click', () => {
      useSelectorMock.mockImplementationOnce((s) => s({ userReducer: { user: mockUser } }));
      useLocationMock.mockReturnValueOnce({
        location: {
          pathname: '/mood-entry',
        },
      });
      const w = shallow(<SiteHeader />);
      w.find('DropdownItem').at(1).simulate('click');
      // TODO: assert if changeLanguage was triggered.
    });

    test('triggers logout call', () => {
      useSelectorMock.mockImplementationOnce((s) => s({ userReducer: { user: mockUser } }));
      useLocationMock.mockReturnValueOnce({
        location: {
          pathname: '/mood-entry',
        },
      });
      const w = shallow(<SiteHeader />);
      w.find('DropdownItem').at(2).simulate('click');
      expect(mockDispatch).toHaveBeenCalledWith({
        type: UserActions.LOGOUT_USER,
      });
      // TODO: assert if changeLanguage was triggered.
    });
  });
});
