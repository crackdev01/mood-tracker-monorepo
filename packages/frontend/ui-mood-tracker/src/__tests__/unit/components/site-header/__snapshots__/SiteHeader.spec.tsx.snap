// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteHeader renders displays header entries for authenticated user 1`] = `
"<Menu pointing={true} secondary={true} className=\\"site-header\\">
  <MenuItem header={true}>
    <Header as=\\"h3\\">
      header
    </Header>
  </MenuItem>
  <MenuItem active={false}>
    <mockConstructor id=\\"mood-entry\\" to=\\"/mood-entry\\">
      moodEntry
    </mockConstructor>
  </MenuItem>
  <MenuItem active={false}>
    <mockConstructor id=\\"statistics\\" to=\\"/statistics\\">
      statistics
    </mockConstructor>
  </MenuItem>
  <MenuMenu position=\\"right\\">
    <MenuItem position=\\"right\\">
      Location permission not provided
    </MenuItem>
    <Dropdown id=\\"toggle-menu\\" item={true} text=\\"\\" additionLabel=\\"Add \\" additionPosition=\\"top\\" closeOnBlur={true} closeOnEscape={true} deburr={false} icon=\\"dropdown\\" minCharacters={1} noResultsMessage=\\"No results found.\\" openOnFocus={true} renderLabel={[Function: renderItemContent]} searchInput=\\"text\\" selectOnBlur={true} selectOnNavigation={true} wrapSelection={true}>
      <DropdownMenu>
        <DropdownItem onClick={[Function: onClick]}>
          menu.actions.location
        </DropdownItem>
        <DropdownItem id=\\"change-language\\" onClick={[Function: updateLanguage]}>
          menu.actions.changeLanguage
        </DropdownItem>
        <DropdownItem id=\\"logout\\" onClick={[Function: logout]}>
          menu.actions.logout
        </DropdownItem>
      </DropdownMenu>
    </Dropdown>
  </MenuMenu>
  <LocationModal displayModal={false} closeModal={[Function: closeLocationModal]} />
</Menu>"
`;

exports[`SiteHeader renders only displays header for unauthenticated user 1`] = `
"<Menu pointing={true} secondary={true} className=\\"site-header\\">
  <MenuItem header={true}>
    <Header as=\\"h3\\">
      header
    </Header>
  </MenuItem>
</Menu>"
`;
